src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -438,15 +451,57 @@ void ReadFromParallelRemoteReplicasStep::initializePipeline(QueryPipelineBuilder
         shuffled_pool = shard.pool->getShuffledPools(current_settings, priority_func);
     }
 
-    for (size_t i=0; i < max_replicas_to_use; ++i)
+    std::vector<ConnectionPoolPtr> pools_to_use;
+    pools_to_use.reserve(shuffled_pool.size());
+    for (const auto & pool : shuffled_pool)
     {
-        IConnections::ReplicaInfo replica_info
+        if (exclude_local_replica)
         {
+            const auto & hostname = pool.pool->getHost();
+            auto it = std::find_if(
+                begin(shard.local_addresses),
+                end(shard.local_addresses),
+                [&hostname](const Cluster::Address & local_addr) { return hostname == local_addr.host_name; });
+            if (it == shard.local_addresses.end())
+                pools_to_use.push_back(pool.pool);
+        }
+        else
+        {
+            pools_to_use.push_back(pool.pool);
+        }
+    }
+
+    pools_to_use.resize(std::min(pools_to_use.size(), max_replicas_to_use));
+    // if local plan is used for local replica, we should exclude one remote replica
+    if (exclude_local_replica && !pools_to_use.empty())
+        pools_to_use.resize(max_replicas_to_use - 1);
+
+    LOG_DEBUG(
+        getLogger("ReadFromParallelRemoteReplicasStep"),
+        "Number of pools to use is {}. Originally {}",
+        pools_to_use.size(),
+        shuffled_pool.size());
+
+    if (pools_to_use.empty())
+        return;
+
+    std::vector<std::string_view> addresses;
+    addresses.reserve(pools_to_use.size());
+    for (const auto & pool : pools_to_use)
+        addresses.emplace_back(pool->getAddress());
+    LOG_DEBUG(getLogger("ReadFromParallelRemoteReplicasStep"), "Addresses to use: {}", fmt::join(addresses, ", "));
+
+    /// when using local plan for local replica, 0 is assigned to local replica as replica num, - in this case, starting from 1 here
+    size_t replica_num = (exclude_local_replica ? 1 : 0);
+    for (const auto & pool : pools_to_use)
+    {
+        IConnections::ReplicaInfo replica_info{
             /// we should use this number specifically because efficiency of data distribution by consistent hash depends on it.
-            .number_of_current_replica = i,
+            .number_of_current_replica = replica_num,
```

Comment:
Sorry, I didn't get it. What is not right here?

Reply:
i got an impression that the old problem revived: now local replica will always be the first replica. and since local replica will always be different depending on which cluster node received a query we will loose consistency of replica numbers => consistency of granules distribution by hash

Reply:
Addressed here [842b51c](https://github.com/ClickHouse/ClickHouse/pull/64448/commits/842b51c782f0cc5cf9bd7b695c466933310a7e6c)